dali/pipeline/data/tensor_vector.cc

```diff
@@ -12,268 +12,494 @@
+#include "dali/core/tensor_shape.h"
+#include "dali/pipeline/data/backend.h"
+#include "dali/pipeline/data/types.h"
 
 namespace dali {
 
+namespace {
+
+/**
+ * @brief Check if both shared pointers have the same managed pointer (not the one returned by
+ * .get())
+ */
+bool same_owner(const std::shared_ptr<void> &x, const std::shared_ptr<void> &y) {
+  if (x.owner_before(y) || y.owner_before(x))
+    return false;
+  return true;
+}
+
+
+bool same_owner(const std::weak_ptr<void> &x, const std::shared_ptr<void> &y) {
+  if (x.owner_before(y) || y.owner_before(x))
+    return false;
+  return true;
+}
+
+
+// TODO(klecki): move this to the class?
+TensorShape<> empty_shape(int dim) {
+  TensorShape<> result;
+  result.resize(dim);
+  for (auto &elem : result) {
+    elem = 0;
+  }
+  return result;
+}
+
+
+/**
+ * @defgroup copy_impl Helper code for copying batches
+ * The functions used as scaffolding for the synchronization of order for source and destination
+ * buffers. Generic copy between contiguous and noncontiguous batches that assumes that the
+ * buffers are already resized.
+ *
+ * @{
+ */
+/**
+ * @brief Pick the order for Copy to be run on.
+ *
+ * The copy ordering can be:
+ * - explict, as specified in `order`
+ * - the one from `src_order`, if set
+ * - the one from `dst_order`
+ * @return copy_order - order on which we will do the copy
+ */
+AccessOrder CopySyncBefore(AccessOrder dst_order, AccessOrder src_order, AccessOrder order) {
+  if (!order)
+    order = src_order ? src_order : dst_order;
+
+  // Wait on the order on which we will run the copy for the work to finish on the dst
+  order.wait(dst_order);
+
+  return order;
+}
+
+
+/**
+ * @brief Wait for the reallocation to happen in the copy order, so we can actually proceed.
+ */
+void CopySyncResize(AccessOrder dst_order, AccessOrder copy_order) {
+  copy_order.wait(dst_order);
+}
+
+
+/**
+ * @brief Wait for the copy to finish in the order of the dst buffer.
+ */
+void CopySyncAfter(AccessOrder dst_order, AccessOrder copy_order) {
+  dst_order.wait(copy_order);
+}
+
+
+template <typename DstBackend, typename SrcBackend, template <typename> typename DstBatch,
+          template <typename> typename SrcBatch>
+void CopySamplewiseImpl(DstBatch<DstBackend> &dst, const SrcBatch<SrcBackend> &src,
+                        const TypeInfo &type_info, AccessOrder order = {},
+                        bool use_copy_kernel = false) {
+  auto num_samples = src.num_samples();
+  SmallVector<const void *, 256> srcs;
+  srcs.reserve(num_samples);
+  SmallVector<void *, 256> dsts;
+  dsts.reserve(num_samples);
+  SmallVector<Index, 256> sizes;
+  sizes.reserve(num_samples);
+  for (int i = 0; i < num_samples; i++) {
+    dsts.emplace_back(dst.raw_mutable_tensor(i));
+    srcs.emplace_back(src.raw_tensor(i));
+    sizes.emplace_back(src.shape()[i].num_elements());
+  }
+
+  type_info.template Copy<SrcBackend, DstBackend>(dsts.data(), srcs.data(), sizes.data(),
```

Comment:
```suggestion
  type_info.Copy<SrcBackend, DstBackend>(dsts.data(), srcs.data(), sizes.data(),
```
type_info is not a dependent name, so Copy shouldn't need a template disambiguation.